docs/resources/server/loottables/index.md

```diff
@@ -0,0 +1,322 @@
+
+```java
+public static final DeferredRegister<LootItemConditionType> LOOT_CONDITION_TYPES =
+        DeferredRegister.create(Registries.LOOT_CONDITION_TYPE, ExampleMod.MOD_ID);
+
+public static final Supplier<LootItemConditionType> MIN_XP_LEVEL =
+        LOOT_CONDITION_TYPES.register("min_xp_level", () -> HasXpLevelCondition.TYPE);
+```
+
+## Custom Loot Functions
+
+Similar to loot conditions, we can also add loot functions by registering a corresponding type object to their [registry], which is in turn created with a [codec]. `LootItemFunction` extends `BiFunction<ItemStack, LootContext, ItemStack>`, so what we want is to use the existing item stack and the loot context to return a new, modified item stack.
+
+Almost all loot functions don't directly extend `LootItemFunction`, but extend `LootItemConditionalFunction` instead. This class has built-in functionality for applying loot conditions to the function - the function is only applied if the loot conditions apply. For the sake of example, let's apply a random enchantment with a specified level to the item:
+
+```java
+// Code adapted from vanilla's EnchantRandomlyFunction class.
+// LootItemConditionalFunction is an abstract class, not an interface, so we cannot use a record here.
+public class RandomEnchantmentWithLevelFunction extends LootItemConditionalFunction {
+    // Our context: an optional list of enchantments, and a level.
+    private final Optional<HolderSet<Enchantment>> enchantments;
+    private final int level;
+    // A codec for the enchantment registry.
+    private static final Codec<HolderSet<Enchantment>> ENCHANTMENT_SET_CODEC = BuiltInRegistries.ENCHANTMENT
+            .holderByNameCodec()
+            .listOf()
+            .xmap(HolderSet::direct, e -> e.stream().toList());
+    // Our codec.
+    public static final Codec<RandomEnchantmentWithLevelFunction> CODEC =
+            // #commonFields adds the conditions field.
+            RecordCodecBuilder.create(inst -> commonFields(inst).and(inst.group(
+                    ExtraCodecs.strictOptionalField(ENCHANTMENT_SET_CODEC, "enchantments").forGetter(inst -> inst.enchantments),
+                    Codec.INT.fieldOf("level").forGetter(inst -> inst.level)
+            ).apply(inst, RandomEnchantmentWithLevelFunction::new));
+    // Our loot function type.
+    public static final LootItemFunctionType TYPE = new LootItemFunctionType(CODEC);
+    
+    RandomEnchantmentWithLevelFunction(List<LootItemCondition> conditions, Optional<HolderSet<Enchantment>> enchantments, int level) {
+        super(conditions);
+        this.enchantments = enchantments;
+        this.level = level;
+    }
+    
+    // Return our loot function type here.
+    @Override
+    public LootItemFunctionType getType() {
+        return TYPE;
+    }
+    
+    // Run our enchantment application logic. Most of this is copied from EnchantRandomlyFunction#run.
+    @Override
+    public ItemStack run(ItemStack stack, LootContext context) {
```

Comment:
Definitely needs a second pass because of 1.21.

Reply:
Applied the 1.21 changes.